File: android/src/main/java/com/otasdk/HashVerifier.kt
package com.otasdk

import java.io.File
import java.io.FileInputStream
import java.security.KeyFactory
import java.security.MessageDigest
import java.security.Signature
import java.security.spec.X509EncodedKeySpec
import android.util.Base64

/**
 * Computes and verifies SHA-256 hashes AND ECDSA P-256 signatures of bundle files.
 * Uses Java's built-in security APIs — no third-party crypto dependency.
 */
object HashVerifier {

    /**
     * Compute SHA-256 of a file and return hex string.
     * Reads in 8KB chunks — handles large bundles without OOM.
     */
    fun computeFileHash(file: File): String {
        val digest = MessageDigest.getInstance("SHA-256")
        FileInputStream(file).use { fis ->
            val buffer = ByteArray(8 * 1024)
            var bytesRead: Int
            while (fis.read(buffer).also { bytesRead = it } != -1) {
                digest.update(buffer, 0, bytesRead)
            }
        }
        return digest.digest().joinToString("") { "%02x".format(it) }
    }

    /**
     * Verify that a file's hash matches the expected value.
     * Uses a constant-time comparison to prevent timing attacks.
     */
    fun verifyFile(file: File, expectedHash: String): Boolean {
        if (expectedHash.isBlank()) return false
        val actual = computeFileHash(file)
        return constantTimeEquals(actual, expectedHash.lowercase())
    }

    /**
     * Verify an ECDSA P-256 signature over the raw bytes of a file.
     *
     * @param file          The file whose bytes were signed (the bundle ZIP).
     * @param signatureB64  Base-64-encoded DER signature produced by the CI signing step.
     * @param publicKeyPem  PEM-encoded SPKI public key stored on the server per-app.
     * @return true if the signature is valid, false otherwise.
     */
    fun verifyEcdsaSignature(file: File, signatureB64: String, publicKeyPem: String): Boolean {
        return try {
            // Strip PEM header/footer and decode
            val stripped = publicKeyPem
                .replace("-----BEGIN PUBLIC KEY-----", "")
                .replace("-----END PUBLIC KEY-----", "")
                .replace("\\s".toRegex(), "")
            val keyBytes = Base64.decode(stripped, Base64.DEFAULT)

            val keySpec = X509EncodedKeySpec(keyBytes)
            val publicKey = KeyFactory.getInstance("EC").generatePublic(keySpec)

            val sig = Signature.getInstance("SHA256withECDSA")
            sig.initVerify(publicKey)

            FileInputStream(file).use { fis ->
                val buffer = ByteArray(8 * 1024)
                var bytesRead: Int
                while (fis.read(buffer).also { bytesRead = it } != -1) {
                    sig.update(buffer, 0, bytesRead)
                }
            }

            val sigBytes = Base64.decode(signatureB64, Base64.DEFAULT)
            sig.verify(sigBytes)
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Constant-time string comparison — prevents timing side-channels.
     */
    private fun constantTimeEquals(a: String, b: String): Boolean {
        if (a.length != b.length) return false
        var result = 0
        for (i in a.indices) {
            result = result or (a[i].code xor b[i].code)
        }
        return result == 0
    }
}

File: ios/HashVerifier.swift
import Foundation
import CommonCrypto
import Security

/// SHA-256 hashing, ECDSA signature verification — all using Apple's built-in APIs.
/// No third-party dependencies — ships with every iOS device.
enum HashVerifier {

    /// Compute SHA-256 of a file. Reads in 64KB chunks — handles large bundles.
    static func computeFileHash(at url: URL) throws -> String {
        let file = try FileHandle(forReadingFrom: url)
        defer { file.closeFile() }

        var context = CC_SHA256_CTX()
        CC_SHA256_Init(&context)

        let chunkSize = 64 * 1024
        while true {
            let chunk = file.readData(ofLength: chunkSize)
            if chunk.isEmpty { break }
            chunk.withUnsafeBytes { ptr in
                _ = CC_SHA256_Update(&context, ptr.baseAddress, CC_LONG(chunk.count))
            }
        }

        var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        CC_SHA256_Final(&digest, &context)
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    /// Compute SHA-256 of in-memory data and return a hex string.
    /// Used to verify each file reconstructed from a delta patch.
    static func sha256Hex(_ data: Data) -> String {
        var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        data.withUnsafeBytes { ptr in
            _ = CC_SHA256(ptr.baseAddress, CC_LONG(data.count), &digest)
        }
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    /// Verify a file's hash using constant-time comparison.
    static func verifyFile(at url: URL, expectedHash: String) -> Bool {
        guard !expectedHash.isEmpty else { return false }
        guard let actual = try? computeFileHash(at: url) else { return false }
        return constantTimeEqual(actual, expectedHash.lowercased())
    }

    // MARK: - ECDSA P-256 Signature Verification

    /// Verify an ECDSA-SHA256 P-256 signature over the raw bytes of a file.
    ///
    /// - Parameters:
    ///   - url:           URL of the downloaded bundle ZIP.
    ///   - signatureB64:  Base-64 DER signature produced by the CI signing step.
    ///   - publicKeyPem:  PEM SPKI public key embedded in the app at build time.
    /// - Returns: `true` if the signature is valid, `false` otherwise.
    static func verifyEcdsaSignature(at url: URL,
                                     signatureB64: String,
                                     publicKeyPem: String) -> Bool {
        guard let sigData = Data(base64Encoded: signatureB64) else { return false }

        // Strip PEM header/footer and decode DER
        let stripped = publicKeyPem
            .replacingOccurrences(of: "-----BEGIN PUBLIC KEY-----", with: "")
            .replacingOccurrences(of: "-----END PUBLIC KEY-----", with: "")
            .replacingOccurrences(of: "\n", with: "")
            .replacingOccurrences(of: "\r", with: "")
            .trimmingCharacters(in: .whitespaces)
        guard let derData = Data(base64Encoded: stripped) else { return false }

        // Import the public key using SecKey
        let attributes: [String: Any] = [
            kSecAttrKeyType as String:       kSecAttrKeyTypeEC,
            kSecAttrKeyClass as String:      kSecAttrKeyClassPublic,
            kSecAttrKeySizeInBits as String: 256,
        ]
        var error: Unmanaged<CFError>?
        guard let publicKey = SecKeyCreateWithData(derData as CFData, attributes as CFDictionary, &error)
        else { return false }

        // Read the file into memory
        guard let fileData = try? Data(contentsOf: url) else { return false }

        // Verify with SecKeyVerifySignature (SHA256 is implicit for ecdsaSignatureMessageX962SHA256)
        let algorithm = SecKeyAlgorithm.ecdsaSignatureMessageX962SHA256
        guard SecKeyIsAlgorithmSupported(publicKey, .verify, algorithm) else { return false }

        return SecKeyVerifySignature(
            publicKey,
            algorithm,
            fileData as CFData,
            sigData as CFData,
            &error
        )
    }

    // MARK: - Helpers

    /// Constant-time string comparison — prevents timing side-channels.
    private static func constantTimeEqual(_ a: String, _ b: String) -> Bool {
        guard a.count == b.count else { return false }
        var result: UInt8 = 0
        zip(a.utf8, b.utf8).forEach { result |= $0 ^ $1 }
        return result == 0
    }
}
